tools/autograd/gen_variable_type.py

```diff
@@ -592,19 +646,36 @@ def gen_variable_type_shard(out, aten_declarations, template_path, suffix, heade
             profiled_wrapper_registrations.append(UNBOXEDONLY_WRAPPER_REGISTRATION.substitute(
                 declaration, class_type='ProfiledType'))
 
+        # Emit TraceType code
+        if not declaration['manual_kernel_registration']:
```

Comment:
cc @smessmer I want to reiterate that it was a mistake to tie `manual_kernel_registration` to control BOTH whether or not we generate registrations for CPUType/TypeDefault, and whether or not we generate Variable/Profiling/Tracing wrappers here. I *absolutely* might want to do a manual registration on CPU/CUDA side, but still get tracing automatically generated for me.

Another thing that I find a bit perplexing is that the logic here doesn't line up with the profiling logic. Profiling appears to be generated unconditionally, even if we specified manual kernel registration. What's going on there?

Reply:
In my initial revision I followed the profiling logic - generating unconditionally, but later I found that some manual registered ops do NOT have tracing logic in it, e.g.: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/VariableTypeManual.cpp#L122; generating a tracing kernel for them unconditionally changes semantics.

I think for Profiling the expectation is a little bit relaxed - it doesn't have to be 100% aligned with old behavior (and it doesn't for various corner cases I handled in next diff).

I think @ilia-cher had plan to move profiling into dispatcher? so I didn't change it to avoid conflict.

If not `manual_kernel_registration` then we will need some other blacklist mechanism to skip these ops if we want to keep them from tracing, or it does not matter to trace them?

Reply:
OK, this answers my second question satisfactorily.  I agree as you've written it is semantics preserving and you should keep it as is. (I think my first point still stands but that's a matter for another day.)